rollup/cmd/gas_oracle/app/app.go

```diff
@@ -106,9 +97,31 @@ func action(ctx *cli.Context) error {
 		}
 	})
 
-	// Start l1relayer process
-	go utils.Loop(subCtx, 10*time.Second, l1relayer.ProcessGasPriceOracle)
-	go utils.Loop(subCtx, 2*time.Second, l2relayer.ProcessGasPriceOracle)
+	// Goroutines that may send transactions periodically.
+	go utils.Loop(loopCtx, 10*time.Second, l1relayer.ProcessGasPriceOracle)
+	go utils.Loop(loopCtx, 2*time.Second, l2relayer.ProcessGasPriceOracle)
+
+	defer func() {
+		log.Info("Graceful shutdown initiated")
+
+		// Prevent new transactions by canceling the loop context.
+		loopCancel()
+
+		// Close relayers to ensure all pending transactions are processed.
+		// This includes any in-flight transactions that have not yet been confirmed.
+		l1relayer.Close()
+		l2relayer.Close()
+
+		// Halt confirmation signal handling by canceling the instance context.
```

Comment:
What do you mean by this comment?

Reply:
It's redundant right now, add it here for completeness. it is intended to end handling confirmations like here:
https://github.com/scroll-tech/scroll/blob/v4.3.50/rollup/internal/controller/relayer/l1_relayer.go#L142
https://github.com/scroll-tech/scroll/blob/v4.3.50/rollup/internal/controller/relayer/l2_relayer.go#L697

Reply:
I just find the phrasing "halt confirmation signal" confusing. I guess it means "Trigger relayer graceful shutdown".

Reply:
actually the graceful shutdown is triggered by:
```
		// Close relayers to ensure all pending transactions are processed.
		// This includes any in-flight transactions that have not yet been confirmed.
		l1relayer.Close()
		l2relayer.Close()
```

Reply:
rephrase the comment in [8d94bbe](https://github.com/scroll-tech/scroll/pull/1069/commits/8d94bbe43eef87b1f11f8d38764cb8bccd27f2de)

Reply:
The graceful shutdown is implemented by the following steps:
1. Stop monitoring new blocks and sending new transactions.
2. Invoke `relayer.Stop()` and wait for `sender.pendingTxs` to be cleared.
3. Cease receiving confirmation signals from the sender.